lib/SILGen/SILGenApply.cpp

```diff
@@ -2933,7 +2933,39 @@ static void emitDelayedArguments(SILGenFunction &SGF,
   }
 }
 
-static Expr *findStorageReferenceExprForBorrow(Expr *e) {
+namespace {
+/// Container to hold the result of a search for the storage reference
+/// when determining to emit a borrow.
+struct StorageRefResult {
+  // The direct storage reference.
+  Expr *storageRef;
+  // The root of the expression that accesses the storage in \c storageRef.
+  Expr *transitiveRoot;
+
+  // Represents an empty result
+  StorageRefResult() : storageRef(nullptr), transitiveRoot(nullptr) {}
+  bool isEmpty() const { return transitiveRoot == nullptr; }
+  operator bool() const { return !isEmpty(); }
+
+  StorageRefResult(Expr *storageRef, Expr *transitiveRoot)
+      : storageRef(storageRef), transitiveRoot(transitiveRoot) {
+    assert(storageRef && transitiveRoot && "use the zero-arg init for empty");
+  }
+
+  // Initializes a storage reference where the base matches the ref.
+  StorageRefResult(Expr *storageRef)
+      : StorageRefResult(storageRef, storageRef) {}
+
+  StorageRefResult withTransitiveRoot(Expr *newRoot) {
+    return StorageRefResult(storageRef, newRoot);
+  }
+
+  // By default, this result is the transitive root.
+  operator Expr *() const { return transitiveRoot; }
```

Comment:
Can you turn this into a method with a name to make it more explicit? Maybe getTransitiveRoot()?